### PR TITLE
Dashboard: Fixes save drawer always comparing changes against first loaded version

### LINKED ISSUE #76448: CRITICAL BUG: TypeError: e._d.getTime is not a function | results in crash when saving Grafana Cloud dashboard
### UPDATE
As per my latest post below, this is not related to Athena or any other specific data source. It appears to be a general issue related to saving dashboards with specific time periods specified - even applying to empty dashboards.

This is a critical bug because users will loose any progress they've made when they attempt to save their changes.

-----------

### What happened?

We are creating an Amazon Athena dashboard in Grafana Cloud. The dashboard template is attached below:

[athena-dashboard.json](https://github.com/grafana/grafana/files/12882429/athena-dashboard.json)

When I work with the dashboard, everything works as expected. I can also save my dashboard updates when I select 'All' in my `signal` dropdown. However, if I select specific signals in the drop down the dashboard crashes upon saving. I am also able to reproduce this if I select 'All' in `signal` and in `aggregation`. I do not know what causes this issue.

The video below shows the behavior and I've pasted both the error output from the browser window and console:

**Video**

https://github.com/grafana/grafana/assets/26184427/a29fc9bd-b34c-4a0e-86d2-ac5b0c11459e

**Browser window error**
```
An unexpected error happened
Details
TypeError: e._d.getTime is not a function

    at Qe (https://grafana-assets.grafana.net/grafana/10.2.0-61719/public/build/1470.848c23aba69169162826.js:2507:4383)
    at div
    at div
    at div
    at Z (https://grafana-assets.grafana.net/grafana/10.2.0-61719/public/build/7894.18f6b3bac734822c76f9.js:120:79442)
    at s (https://grafana-assets.grafana.net/grafana/10.2.0-61719/public/build/1470.848c23aba69169162826.js:204:30031)
    at div
    at h (https://grafana-assets.grafana.net/grafana/10.2.0-61719/public/build/7894.18f6b3bac734822c76f9.js:642:78922)
    at div
    at O (https://grafana-assets.grafana.net/grafana/10.2.0-61719/public/build/7894.18f6b3bac734822c76f9.js:39:38163)
    at div
    at ke (https://grafana-assets.grafana.net/grafana/10.2.0-61719/public/build/7894.18f6b3bac734822c76f9.js:39:48888)
    at https://grafana-assets.grafana.net/grafana/10.2.0-61719/public/build/7894.18f6b3bac734822c76f9.js:39:49199
    at div
    at R (https://grafana-assets.grafana.net/grafana/10.2.0-61719/public/build/7894.18f6b3bac734822c76f9.js:39:39160)
    at https://grafana-assets.grafana.net/grafana/10.2.0-61719/public/build/7894.18f6b3bac734822c76f9.js:26:296
    at P (https://grafana-assets.grafana.net/grafana/10.2.0-61719/public/build/7894.18f6b3bac734822c76f9.js:39:42581)
    at u (https://grafana-assets.grafana.net/grafana/10.2.0-61719/public/build/1470.848c23aba69169162826.js:204:112169)
    at at (https://grafana-assets.grafana.net/grafana/10.2.0-61719/public/build/1470.848c23aba69169162826.js:2518:21)
    at r
    at l (https://grafana-assets.grafana.net/grafana/10.2.0-61719/public/build/1470.848c23aba69169162826.js:224:13290)
    at v (https://grafana-assets.grafana.net/grafana/10.2.0-61719/public/build/7894.18f6b3bac734822c76f9.js:37:13808)
    at d (https://grafana-assets.grafana.net/grafana/10.2.0-61719/public/build/7894.18f6b3bac734822c76f9.js:642:177319)
    at c (https://grafana-assets.grafana.net/grafana/10.2.0-61719/public/build/1470.848c23aba69169162826.js:1076:20362)
    at o (https://grafana-assets.grafana.net/grafana/10.2.0-61719/public/build/1470.848c23aba69169162826.js:204:119127)
    at m (https://grafana-assets.grafana.net/grafana/10.2.0-61719/public/build/1470.848c23aba69169162826.js:204:119636)
    at y (https://grafana-assets.grafana.net/grafana/10.2.0-61719/public/build/7894.18f6b3bac734822c76f9.js:181:94044)
    at ni (https://grafana-assets.grafana.net/grafana/10.2.0-61719/public/build/1470.848c23aba69169162826.js:384:127)
```

I am using Chrome and my console output is below:

```
module.js?_cache=0.6.7:452 TypeError: e._d.getTime is not a function
    at new ve (moment.js:262:58)
    at sa (moment.js:2968:20)
    at dn (moment.js:2943:44)
    at na (moment.js:3037:16)
    at A (moment.js:3041:16)
    at a (moment.js:16:29)
    at g (moment_wrapper.ts:117:10)
    at U.hasTimeChanged (DashboardModel.ts:1095:38)
    at SaveDashboardForm.tsx:39:50
    at Object.useMemo (react-dom.production.min.js:184:189)

```

### What did you expect to happen?

The dashboard should save as expected without crashing

### Did this work before?

I do not know

### How do we reproduce it?

I am unable to provide a directly reproducible example

### Is the bug inside a dashboard panel?

_No response_

### Environment (with versions)?

Grafana: Grafana v10.2.0-61719 (165b4c3f66) | Cloud

OS: Windows 10 64 bit
Browser: Chrome

### Grafana platform?

None

### Datasource(s)?

_No response_

### PR DESCRIPTION
**Pull request by torkelo**
Bug introduced in https://github.com/grafana/grafana/pull/73881 

Fixes https://github.com/grafana/grafana/issues/76448

When you save a dashboard the version we compare against in the drawer was never updated. 

Though this was going to be a quick fix but o shit the typing issues we have with using DashboardModel and schema (save model) Dashboard interchangeably / incorrectly. Made me want to go back to bed and cry. 

I tried to fix it so that getSaveModelClone returns the save model (Schema) type but get like 80 type errors, some tricky ones. So I ended up introducing a new function and renaming the old function, this is mainly to keep this PR smaller to fix those types of errors separately. The runtime behavior of both functions should be identical. The only difference is that for some inexplicable reason, getSaveModelClone adds a getVariables function on the save model!

But even this fix required A LOT of fixes.

### CODE REVIEW

File: public/app/features/dashboard/state/DashboardModel.ts [hunk 1]
```diff
@@ -577,6 +592,8 @@ export class DashboardModel implements TimeModel {
       this.panelInEdit.hasSavedPanelEditChange = this.panelInEdit.configRev > 0;
       this.panelInEdit.configRev = 0;
     }
+
+    this.originalDashboard = this.getSaveModelClone();
```
Comment (dprokop) [inline public/app/features/dashboard/state/DashboardModel.ts:596]: For anyone else reviewing - this is the actual fix ;)

**Review by dprokop — APPROVED** [review]
Looks good. Tested the behavior and now it works as it should.

File: public/app/features/dashboard/state/DashboardModel.ts [hunk 2]
```diff
@@ -368,36 +368,34 @@ export class DashboardModel implements TimeModel {
...
 
-        if (current.type === 'adhoc') {
-          current.filters = original.filters;
-        } else {
-          current.current = original.current;
+          if (variable.type === 'adhoc') {
+            variableSaveModel.filters = original.filters;
+          } else {
+            variableSaveModel.current = original.current;
+            variableSaveModel.options = original.options;
```
Comment (dprokop) [inline public/app/features/dashboard/state/DashboardModel.ts:392]: Do we need to save the options? They are fetched anyways on load and https://github.com/grafana/grafana/pull/76403 should make sure the custom one is handled as well.

Comment (dprokop) [reply public/app/features/dashboard/state/DashboardModel.ts:392]: aha, ok we need not to trigger the changes detection for every single dashboard after this is merged.

**Review by axelavargas — APPROVED** [review]
Hey @torkelo 👋🏾 , I have tested the latest changes against scenarios using variables, links, and panel changes. It works as expected 🙌🏾 . 
Codewise I am a bit lost with this implementation, so I can't give feedback there.

**Comment by MatinF** [comment]
It looks like #76448 is to be fixed by this commit - but it does not yet seem to affect Grafana Cloud. Is it correct to assume this has not yet been rolled out?

**Pull request merged**